DOCSP-41981: Change streams (#113)
mongodb/docs-php-library: source/read/change-streams.txt

.. _php-change-streams:

====================
Monitor Data Changes
====================

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

.. facet::
   :name: genre
   :values: reference

.. meta::
   :keywords: watch, code example

Overview
--------

In this guide, you can learn how to use a **change stream** to monitor real-time
changes to your data. A change stream is a {+mdb-server+} feature that
allows your application to subscribe to data changes on a collection, database,
or deployment.

> When using the {+php-library+}, you can instantiate a ``MongoDB\ChangeStream`` to
> monitor data changes.

Reviewer: ```suggestion
When using the {+php-library+}, you can create an instance of ``MongoDB\ChangeStream`` to
monitor data changes, such as updates, insertions, and deletions.
```

stream monitors. You can call the ``watch()`` method on instances of the following
classes:

- ``MongoDB\Client``: Monitor all changes in the MongoDB deployment
- ``MongoDB\Database``: Monitor changes in all collections in the database
- ``MongoDB\Collection``: Monitor changes in the collection

The following example opens a change stream on the ``restaurants`` collection
and outputs changes as they occur:

.. literalinclude:: /includes/read/change-streams.php
   :start-after: start-open-change-stream
   :end-before: end-open-change-stream
   :language: php
   :dedent:

To begin watching for changes, run the preceding code. Then, in a separate
application or shell, modify the ``restaurants`` collection. The following
example updates a document that has a ``name`` field value of ``'Blarney Castle'``:

.. _php-change-stream-update:

.. literalinclude:: /includes/read/change-streams.php
   :start-after: start-update-for-change-stream
   :end-before: end-update-for-change-stream
   :language: php
   :dedent:

When you update the collection, the change stream application prints the change
as it occurs. The printed change event resembles the following output:

> .. code-block:: bash

Reviewer: S: change the type to `none` instead of bash, as this is output not a command. I think `console` might work too

   :copyable: false

   {"_id":{"_data":"..."},"operationType":"update","clusterTime":{"$timestamp":{...},
   "wallTime":{"$date":...},"ns":{"db":"sample_restaurants","coll":"restaurants"},
   "documentKey":{"_id":{"$oid":"..."}},"updateDescription":{"updatedFields":
   {"cuisine":"Irish"},"removedFields":[],"truncatedArrays":[]}}}

Modify the Change Stream Output
-------------------------------

To modify the change stream output, you can pass pipeline stages in an array as a
parameter to the ``watch()`` method. You can include the following stages in the
array:

- ``$addFields`` or ``$set``: Adds new fields to documents
- ``$match``: Filters the documents
- ``$project``: Projects a subset of the document fields
- ``$replaceWith`` or ``$replaceRoot``: Replaces the input document with the
  specified document
- ``$redact``: Restricts the contents of the documents
- ``$unset``: Removes fields from documents

> The following passes a pipeline that includes the ``$match`` stage to the ``watch()``
> method. This instructs the ``watch()`` method to output only update operations:

Reviewer: ```suggestion
The following passes a pipeline that includes the ``$match`` stage to the ``watch()``
method. This instructs the ``watch()`` method to output events only when update operations occur:
```

.. literalinclude:: /includes/read/change-streams.php
   :start-after: start-change-stream-pipeline
   :end-before: end-change-stream-pipeline
   :language: php
   :dedent:

Modify ``watch()`` Behavior
---------------------------

To modify the behavior of the ``watch()`` method, you can pass an options array

> as a parameter to ``watch()``. The following table describes some options you

Reviewer: ```suggestion
as a parameter to ``watch()``. The following table describes useful options you
```

can set in the array:

.. list-table::
   :widths: 30 70
   :header-rows: 1

   * - Option
     - Description

   * - ``fullDocument``
     - | Specifies whether to show the full document after the change, rather
         than showing only the changes made to the document. To learn more about

>          this option, see :ref:`php-change-stream-pre-post-image`.

Reviewer: ```suggestion
         this option, see the :ref:`php-change-stream-pre-post-image` section of this guide.
```

   * - ``fullDocumentBeforeChange``
     - | Specifies whether to show the full document as it was before the change, rather
         than showing only the changes made to the document. To learn more about
         this option, see :ref:`php-change-stream-pre-post-image`.
    
   * - ``resumeAfter``
     - | Instructs ``watch()`` to resume returning changes after the
         operation specified in the resume token.
       | Each change stream event document includes a resume token as the ``_id``
         field. Pass the entire ``_id`` field of the change event document that
         represents the operation you want to resume after.

>        | This option is mutually exclusive with ``startAfter`` and ``startAtOperationTime``.

Reviewer: S: since you mention `startAtOperationTime`, maybe you can include it in this table

one of the following values:

- ``MongoDB\Operation\Watch::FULL_DOCUMENT_UPDATE_LOOKUP``: The change event includes a
  copy of the entire changed document from some time after the change.
- ``MongoDB\Operation\Watch::FULL_DOCUMENT_WHEN_AVAILABLE``: The change event includes
  a post-image of the modified document for change events only if the post-image is available.
- ``MongoDB\Operation\Watch::FULL_DOCUMENT_REQUIRED``: The change event includes a post-image
  of the modified document for change events. If the post-image is not available, the
  {+php-library+} raises an error.

The following example calls the ``watch()`` method on a collection and includes the post-image
of updated documents by setting the ``fullDocument`` option:

.. literalinclude:: /includes/read/change-streams.php
   :start-after: start-change-stream-post-image
   :end-before: end-change-stream-post-image
   :language: php
   :dedent:

With the change stream application running, updating a document in the
``restaurants`` collection by using the :ref:`preceding update example
<php-change-stream-update>` prints a change event resembling the following

> code:

Reviewer: ```suggestion
output:
```

.. _php-change-streams:

====================
Monitor Data Changes
====================

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

.. facet::
   :name: genre
   :values: reference

.. meta::
   :keywords: watch, code example

Overview
--------

In this guide, you can learn how to use a **change stream** to monitor real-time
changes to your data. A change stream is a {+mdb-server+} feature that
allows your application to subscribe to data changes on a collection, database,
or deployment.

When using the {+php-library+}, you can create an instance of ``MongoDB\ChangeStream`` to

> monitor data changes, such as updates, insertions, and deletions.

Reviewer 2: Users never construct a `MongoDB\ChangeStream` object. It'd be more correct to say they can invoke the various `watch()` methods on Client, Database, or Collection objects to obtain a ChangeStream object that they can then iterate to monitor data changes on the logical entity (cluster, database, or collection, respectively).

Sample Data
~~~~~~~~~~~

The examples in this guide use the ``restaurants`` collection in the ``sample_restaurants``
database from the :atlas:`Atlas sample datasets </sample-data>`. To access this collection
from your PHP application, instantiate a ``MongoDB\Client`` that connects to an Atlas cluster

> and assign the following value to your ``collection`` variable:

Reviewer 2: `$client->sample_restaurants->restaurants` constructs a new collection, so it'd be more accurate to say something like "select a collection to assign to the ``collection`` variable" or "instantiate your ``collection`` variable".

Also, since you're referring to a variable by name here, would it make sense to use ``$collection``? That would be consistent with how use a "()" suffix when referring to methods like `watch()` below.

Author: Fixed!

Author: Changed to `$collection`! Do you mean that `$client->sample_restaurants->restaurants` creates a new `restaurants` collection rather than accessing the `restaurants` sample collection? From running the code, it does query the sample data

Reviewer 2: My objection was with the word "value" in "assign the following value" because `$client->sample_restaurants->restaurants` is not actually a value even though it looks like we're accessing PHP properties.

The code here is using the [`Database::__get()`](https://github.com/mongodb/mongo-php-library/blob/1.19.1/src/Database.php#L163) method, which delegates to [`Database::selectCollection()`](https://github.com/mongodb/mongo-php-library/blob/1.19.1/src/Database.php#L559). That method constructs a new `MongoDB\Collection` object each time it is called. To me, "value" implies that we're assigning some fixed value.

If it's kosher to write "assign the following to your `$collection` variable:" I'd be in favor of that. But if "value" needs to stay we can resolve this thread without further changes.

Note: In the PHPLIB docs I've historically used the `Client::selectDatabase()`, `Client::selectCollection()`, and `Database::selectCollection()` methods; however, I understand that you might want to use the shorthand syntax for consistency with other driver/shell docs.

> Do you mean that $client->sample_restaurants->restaurants creates a new restaurants collection rather than accessing the restaurants sample collection?

Creation of the `MongoDB\Collection` object has no impact on the server. We're just initializing the object with strings to identify the database and collection (i.e. namespace). The driver will only create a collection on the server if we explicitly invoke the [`create` command](https://www.mongodb.com/docs/manual/reference/command/create/) or write to a collection that doesn't exist (implicit creation).

The following example opens a change stream on the ``restaurants`` collection
and outputs changes as they occur:

.. literalinclude:: /includes/read/change-streams.php
   :start-after: start-open-change-stream
   :end-before: end-open-change-stream
   :language: php
   :dedent:

To begin watching for changes, run the preceding code. Then, in a separate
application or shell, modify the ``restaurants`` collection. The following
example updates a document that has a ``name`` field value of ``'Blarney Castle'``:

.. _php-change-stream-update:

.. literalinclude:: /includes/read/change-streams.php
   :start-after: start-update-for-change-stream
   :end-before: end-update-for-change-stream
   :language: php
   :dedent:

When you update the collection, the change stream application prints the change
as it occurs. The printed change event resembles the following output:

.. code-block:: none
   :copyable: false

   {"_id":{"_data":"..."},"operationType":"update","clusterTime":{"$timestamp":{...},
   "wallTime":{"$date":...},"ns":{"db":"sample_restaurants","coll":"restaurants"},
   "documentKey":{"_id":{"$oid":"..."}},"updateDescription":{"updatedFields":

>    {"cuisine":"Irish"},"removedFields":[],"truncatedArrays":[]}}}

Reviewer 2: If `json_encode()` is used in the PHP example, I don't think you'll see keys like "$timestamp", "$date", and "$oid" in the output here. The BSON API will need to be used to ensure Extended JSON syntax is produced.

method. This instructs the ``watch()`` method to output events only when update
operations occur:

.. literalinclude:: /includes/read/change-streams.php
   :start-after: start-change-stream-pipeline
   :end-before: end-change-stream-pipeline
   :language: php
   :dedent:

Modify ``watch()`` Behavior
---------------------------

To modify the behavior of the ``watch()`` method, you can pass an options array
as a parameter to ``watch()``. The following table describes useful options you
can set in the array:

.. list-table::
   :widths: 30 70
   :header-rows: 1

   * - Option
     - Description

   * - ``fullDocument``
     - | Specifies whether to show the full document after the change, rather
         than showing only the changes made to the document. To learn more about
         this option, see the :ref:`php-change-stream-pre-post-image` section of this
         guide.

   * - ``fullDocumentBeforeChange``
     - | Specifies whether to show the full document as it was before the change, rather
         than showing only the changes made to the document. To learn more about
         this option, see :ref:`php-change-stream-pre-post-image`.
    

>    * - ``resumeAfter``

Reviewer 2: Note: I assume this change stream tutorial is common to other driver docs, so consider this a more general question beyond the scope of PHPLIB.

Is `resumeAfter` actually relevant to the users? I know it's the original option and used internally by drivers when resuming, but my understanding is that `startAfter` is preferred for users.

Looking at [`db.collection.watch()`](https://www.mongodb.com/docs/manual/reference/method/db.collection.watch/), which documents both options, it looks like `resumeAfter` only differs by its limitation to _not_ allow resuming after an "invalidate" event. Given that, it doesn't seem like a "useful option" to include in this table (although it absolutely belongs in the API reference).

Author: If startAfter is preferred, I'm happy to remove resumeAfter

         operation specified in the resume token.
       | Each change stream event document includes a resume token as the ``_id``
         field. Pass the entire ``_id`` field of the change event document that
         represents the operation you want to resume after.
       | This option is mutually exclusive with ``startAfter`` and ``startAtOperationTime``.

   * - ``startAfter``
     - | Instructs ``watch()`` to start a new change stream after the
         operation specified in the resume token. This field allows notifications to
         resume after an invalidate event.
       | Each change stream event document includes a resume token as the ``_id``
         field. Pass the entire ``_id`` field of the change event document that
         represents the operation you want to resume after.
       | This option is mutually exclusive with ``resumeAfter`` and ``startAtOperationTime``.

   * - ``startAtOperationTime``
     - | Instructs the change stream to only provide changes that occurred at or after
         the specified timestamp. 
       | This option is mutually exclusive with ``startAfter`` and ``resumeAfter``.

>    * - ``typeMap``

Reviewer 2: This also seems like something we can omit from this page if the table is admittedly a subset of "useful options". It isn't demonstrated in any examples, and it's already covered in the API reference.

.. _php-change-stream-pre-post-image:

Include Pre-Images and Post-Images
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. important::

   You can enable pre-images and post-images on collections only if your
   deployment uses MongoDB v6.0 or later.

By default, when you perform an operation on a collection, the
corresponding change event includes only the delta of the fields
modified by that operation. To see the full document before or after a
change, specify the ``fullDocumentBeforeChange`` or the ``fullDocument``
options in an array parameter to ``watch()``.

The **pre-image** is the full version of a document *before* a change. To include the
pre-image in the change stream event, set the ``fullDocumentBeforeChange`` option
to one of the following values:

- ``MongoDB\Operation\Watch::FULL_DOCUMENT_BEFORE_CHANGE_WHEN_AVAILABLE``: The change event includes
  a pre-image of the modified document for change events. If the pre-image is not available, this
  change event field has a ``null`` value.
- ``MongoDB\Operation\Watch::FULL_DOCUMENT_BEFORE_CHANGE_REQUIRED``: The change event includes a pre-image
  of the modified document for change events. If the pre-image is not available, the

>   {+php-library+} raises an error.  

Reviewer 2: Technically, the server raises errors for this and the corresponding post-image case. PHPLIB would just propagate that error to the user. Not sure if this is important to clarify, as this might mislead users into thinking it's a client-side error.

Reviewer 2: Did you mean to change the text here to clarify that PHPLIB will propagate the server-side error?

Author: oops yes I did, fixed now

The **post-image** is the full version of a document *after* a change. To include the
post-image in the change stream event, set the ``fullDocument`` option to
one of the following values:

- ``MongoDB\Operation\Watch::FULL_DOCUMENT_UPDATE_LOOKUP``: The change event includes a
  copy of the entire changed document from some time after the change.
- ``MongoDB\Operation\Watch::FULL_DOCUMENT_WHEN_AVAILABLE``: The change event includes

>   a post-image of the modified document for change events only if the post-image is available.

Reviewer 2: Should this clarify that `null` is used as a filler value, as you do for pre-image?

.. _php-change-streams:

====================
Monitor Data Changes
====================

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

.. facet::
   :name: genre
   :values: reference

.. meta::
   :keywords: watch, code example

Overview
--------

In this guide, you can learn how to use a **change stream** to monitor real-time
changes to your data. A change stream is a {+mdb-server+} feature that
allows your application to subscribe to data changes on a collection, database,
or deployment.

When using the {+php-library+}, you can call the ``watch()`` method to return an
instance of ``MongoDB\ChangeStream``. Then, you can iterate through change events stored 

> in the ``MongoDB\ChangeStream`` instance to monitor data changes, such as updates,

Reviewer 2: Unless you need to include the phrase "change events" here, consider "...iterate through the ``MongoDB\ChangeStream`` instance to..." as a more concise alternative.

I don't think of cursor objects as something that store data, so the original phrasing as odd to me. Ignore this feedback if this is the same language you're using in other language docs, though.

Author: changed!

--------------------

To open a change stream, call the ``watch()`` method. The instance on which you
call the ``watch()`` method determines the scope of events that the change
stream monitors. You can call the ``watch()`` method on instances of the following
classes:

- ``MongoDB\Client``: Monitor all changes in the MongoDB deployment
- ``MongoDB\Database``: Monitor changes in all collections in the database
- ``MongoDB\Collection``: Monitor changes in the collection

The following example opens a change stream on the ``restaurants`` collection
and outputs changes as they occur:

.. literalinclude:: /includes/read/change-streams.php
   :start-after: start-open-change-stream
   :end-before: end-open-change-stream
   :language: php
   :dedent:

To begin watching for changes, run the preceding code. Then, in a separate
application or shell, modify the ``restaurants`` collection. The following
example updates a document that has a ``name`` field value of ``'Blarney Castle'``:

.. _php-change-stream-update:

.. literalinclude:: /includes/read/change-streams.php
   :start-after: start-update-for-change-stream
   :end-before: end-update-for-change-stream
   :language: php
   :dedent:

> When you update the collection, the change stream application prints the change

Reviewer 2: Something that just occurred to me: should this tutorial advise users to run this PHP code through a shell/terminal? Users would likely run into issues trying to execute the `watch()` process through a web server.

I'm curious if there's language from other language tutorials (assuming this was adapted from existing content) we can use.

The most straightforward way to address this might be to consistently refer to "shell process" instead of "application or shell" for all of the code examples (no harm in using that for the update example as well).

Author: We use "application" pretty consistently across all docs, but I agree that it makes sense to suggest running the application from the shell. I updated all mentions of running the application on this page to clarify that

  {+php-library+} raises an error.  

The **post-image** is the full version of a document *after* a change. To include the
post-image in the change stream event, set the ``fullDocument`` option to
one of the following values:

- ``MongoDB\Operation\Watch::FULL_DOCUMENT_UPDATE_LOOKUP``: The change event includes a
  copy of the entire changed document from some time after the change.
- ``MongoDB\Operation\Watch::FULL_DOCUMENT_WHEN_AVAILABLE``: The change event includes
  a post-image of the modified document for change events. If the post-image is not
  available, this change event field has a ``null`` value.
- ``MongoDB\Operation\Watch::FULL_DOCUMENT_REQUIRED``: The change event includes a post-image
  of the modified document for change events. If the post-image is not available, the
  server raises an error.

The following example calls the ``watch()`` method on a collection and includes the post-image
of updated documents by setting the ``fullDocument`` option:

.. literalinclude:: /includes/read/change-streams.php
   :start-after: start-change-stream-post-image
   :end-before: end-change-stream-post-image
   :language: php
   :dedent:

> With the change stream application running, updating a document in the

Reviewer 2: I think "updating" should be "update".

Per my earlier comment about pointing users away from using a web server, the following might be clearer:

> With the change stream application running in a separate shell, update a document...

Author: Added "in a separate shell" - "using" makes more grammatical sense in this case

Reviewer 2: > "updating" makes more grammatical sense in this case

My mistake. I missed that the sentence continued with "prints a change event..." after the `:ref:` link.
